test/gateway/eip20_gateway/progress_unstake.js

```diff
@@ -0,0 +1,527 @@
+        unstakeMessage.unlockSecret,
+      ),
+      'Message on target status must be Declared.',
+    );
+
+  });
+
+  it('should fail for revoked redeem(unstake) message', async function () {
+
+    await gateway.setInboxStatus(
+      unstakeMessage.messageHash,
+      MessageStatusEnum.Revoked,
+    );
+
+    await Utils.expectRevert(
+      gateway.progressUnstake(
+        unstakeMessage.messageHash,
+        unstakeMessage.unlockSecret,
+      ),
+      'Message on target status must be Declared.',
+    );
+
+  });
+
+  it('should fail when the reward amount is greater than the unstake amount',
+    async function () {
+
+    unstakeMessage.gasPrice = new BN(10000);
+    unstakeMessage.gasLimit = new BN(10000);
+
+    await setMessage();
+
+    await gateway.setInboxStatus(
+      unstakeMessage.messageHash,
+      MessageStatusEnum.Declared,
+    );
+
+    await Utils.expectRevert(
+      gateway.progressUnstake(
+        unstakeMessage.messageHash,
+        unstakeMessage.unlockSecret,
+      ),
+      'Reward amount must be less than redeem amount.',
+    );
+
+  });
+
+  it('should return correct "redeemAmount", "unstakeAmount" and ' +
+    '"rewardAmount" when gas price is zero', async function () {
+
+    unstakeMessage.gasPrice = new BN(0);
+
+    await setMessage();
+
+    await gateway.setInboxStatus(
+      unstakeMessage.messageHash,
+      MessageStatusEnum.Declared,
+    );
+
+    let result = await gateway.progressUnstake.call(
+      unstakeMessage.messageHash,
+      unstakeMessage.unlockSecret,
+    );
+
+    assert.strictEqual(
+      unstakeRequest.amount.eq(result.redeemAmount_),
+      true,
+      `Redeem amount ${result.redeemAmount_.toString(10)} must be equal to ${unstakeRequest.amount.toString(10)}`,
+    );
+
+    assert.strictEqual(
+      unstakeRequest.amount.eq(result.unstakeAmount_),
+      true,
+      `Unstake amount ${result.unstakeAmount_.toString(10)} must be equal to ${unstakeRequest.amount.toString(10)}`,
+    );
+
+    assert.strictEqual(
+      result.rewardAmount_.eqn(0),
+      true,
+      `Reward amount ${result.rewardAmount_.toString(10)} must be equal to zero`,
+    );
+
+  });
+
+  it('should return correct "redeemAmount", "unstakeAmount" and ' +
+    '"rewardAmount" when gas price is greater than zero', async function () {
+
+    unstakeMessage.gasPrice = new BN(1);
+    unstakeMessage.gasLimit = new BN(1000000000);
+
+    await setMessage();
+
+    await gateway.setInboxStatus(
+      unstakeMessage.messageHash,
+      MessageStatusEnum.Declared,
+    );
+
+    let tx = await gateway.progressUnstake(
+      unstakeMessage.messageHash,
+      unstakeMessage.unlockSecret,
+    );
+
+    let event = EventDecoder.getEvents(tx, gateway);
+    let eventData = event.UnstakeProgressed;
+
+    let gasUsed = new BN(tx.receipt.gasUsed);
+    let maxReward = gasUsed.mul(unstakeMessage.gasPrice);
+
+    assert.strictEqual(
+      eventData._rewardAmount.lt(maxReward),
```

Comment:
This definitely needs an explanatory comment.

Reply:
Added